### PR TITLE
network: Add support specify the outgoing interface for FDB entries

### PR DESCRIPTION
**Pull request by ssahani**
Closes https://github.com/systemd/systemd/issues/17247

### CODE REVIEW

File: src/network/networkd-fdb.c
Comment (poettering) [inline src/network/networkd-fdb.c:455]: this doesn't look right: parsing this at the time the .network files are read might be too late. this should store the ifname as text and resolve when the fdb record is actually installed, so that this works for all network interfaces created between the time networkd reads the .network files and then finally acts on them.

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/network/networkd-fdb.c
Comment (yuwata) [reply src/network/networkd-fdb.c:455]: Yeah, but we already use resolve_ifname() at many config parsers. We need to fix them at some point...

Comment (ssahani) [reply src/network/networkd-fdb.c:455]: How does it fix ? altname or rename

Comment (yuwata) [reply src/network/networkd-fdb.c:455]: Maybe, store the ifname as is in conf parsers, and then resolve it in configure_xxx().

Comment (ssahani) [reply src/network/networkd-fdb.c:455]: Is it the timing ?

Comment (yuwata) [reply src/network/networkd-fdb.c:455]: Yeah, at that time when we parse the .network files, the corresponding netdevs may not exist yet.

Comment (yuwata) [reply src/network/networkd-fdb.c:455]: But to handle such scenario is not simple. Maybe keep this as is, and leave to a later PR is ok at this time, I think. @poettering WDYT?

Comment (ssahani) [reply src/network/networkd-fdb.c:455]: yes it's strange how to handle this.

Comment (poettering) [reply src/network/networkd-fdb.c:455]: if this is already a pre-existing pattern it might be Ok to merge this with the current logic, but this really should be fixed I guess. I don#t think the fix is going to be difficult: it just means storing the string the user specified and then at the time of applying things looking for the right ifindex for it. Given that networkd keeps tracks of all ifaces anyway that should be pretty easy.

File: src/network/networkd-fdb.c
Comment (poettering) [inline src/network/networkd-fdb.c]: This can fail, no? that should be handled somehow. probably cause failure here. at the minimum cause a log message.

hmm, instead of using resolve_ifname() it should be possible to look this up with the data networkd already loaded, no?

i.e. the Manager object's `.links` hashmap field should contain this data. except that the hashmap is indexed by the ifindex rather than the name. But I think it would make a ton of sense to add an adjacent hashmap that keys things by ifname and points to the Link object.

**Review by poettering — CHANGES_REQUESTED** [review]
This looks pretty OK as it stands now, to me, I'll leave it to @yuwata to merge that.

As mentioned, I think the Manager object should be improved so that we don't have to go to rtnl to look up an ifname (or ifaltname) to an ifindex. after all we have all the data in memory anyway for that, but not indexed in any hash table. i.e. we should add thta hash table so that that just works.

File: src/network/networkd-fdb.c
Comment (ssahani) [reply src/network/networkd-fdb.c]: Sorry for late reply. Yes but I think rather keeping extra hashmap we can introduce a function `manager_find_link_by_ifname` to do the same. Please see below.

File: src/network/networkd-manager.c
Comment (poettering) [inline src/network/networkd-manager.c]: hmm, if there are a lot of interfaces this is going to be slow because it takes linear time. And I figure people do have large amounts of ifaces.

i.e. implementing it this way is O(n) time for n being the number of network interfaces the system has. That's why I think we should have an explicit Hashtable in place we can use to look-up the interface name with. That'd bring the usual access time down to O(1).

I think having a function manager_find_link_by_ifname() is totally the way to go, I just think the backing algorithm/data structure should be different...

Also, this should really cover altnames, too.

File: src/network/networkd-fdb.c
Comment (poettering) [inline src/network/networkd-fdb.c]: Don't show `%m`, it makes no sense here, as you have no error code to show

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by ssahani** [comment]
Updated thanks for the review ptal.

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: the names need to be removed again, when the link goes away. And when the link is renamed/altnames added/removed the hash table also needs updating

**Comment by poettering** [comment]
i like where this is going, but the hashtable needs a bit more love to always stay correctly up-to-date, see above

File: src/network/networkd-link.c
Comment (ssahani) [reply src/network/networkd-link.c]: Ahh right. updated thanks.

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: as a safety precaution I think we should use hashmap_remove_value() here. The only difference is that it also takes both key *and* value when removing, and only removes the entry when the key actually points to the indicated value. i.e use 

```c
hashmap_remove_value(link->manager->links_by_name, *n, link);
```

This might be useful in case the hash table ever gets out of sync, because of some bug.

**Comment by poettering** [comment]
looks good, but link_update() needs to be adjusted too, to update the new hash table whenever IFLA_IFNAME or IFLA_ALT_IFNAME change

**Comment by ssahani** [comment]
> looks good, but link_update() needs to be adjusted too, to update the new hash table whenever IFLA_IFNAME or IFLA_ALT_IFNAME change

Yes right  I removed and added them back in case they changed in `link_update` thanks. But I tried to test it adding a altname while `networkd` is running, the kernel is not pushing the changes or may be we need to add some other flag in the monitor. Updated ptal.

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: oom not handled. maybe better move this into link_drop() and link_add() and make the latter fail if this doesn't work?

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: oom not checked.

i guess the negative effect of this is not too bad, hence might make sense, to just log about this and proceed, i.e. `log_warning_errno()`

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/network/networkd-link.c
Comment (ssahani) [reply src/network/networkd-link.c]: OK

**Comment by ssahani** [comment]
Updated ptal thanks !

**Comment by poettering** [comment]
hmm, but the hash table is not updated in link_add()/link_drop either here. What am I missing?

**Comment by ssahani** [comment]
`link_add` calls `link_new` there it gets updated and link_drop it's getting updated https://github.com/systemd/systemd/pull/17248/commits/defbb4a65f7a5eacf2345707ae07fc9790df2297#diff-3530f9d105ef619fb06e6c8b7fff216475b94d37d9be2a1b673648a7d0080925R1749.

File: src/network/networkd-link.c
Comment (yuwata) [inline src/network/networkd-link.c]: This condition is not necessary.

File: src/network/networkd-link.c
Comment (yuwata) [inline src/network/networkd-link.c]: This condition is not necessary.

File: src/network/networkd-link.c
Comment (yuwata) [inline src/network/networkd-link.c]: Why do you use `hashmap_remove_value()`? It seems `hashmap_remove()` is sufficient.

File: src/network/networkd-link.c
Comment (yuwata) [inline src/network/networkd-link.c]: Same here. Whey `hashmap_remove_value()`?

File: src/network/networkd-link.c
Comment (yuwata) [inline src/network/networkd-link.c]: `hashmap_remove()` seems sufficient.

File: src/network/networkd-fdb.c
Comment (yuwata) [inline src/network/networkd-fdb.c]: I think, this should accept alternative names. Please see `ifname_valid_full()`.

File: src/network/networkd-fdb.c
Comment (yuwata) [inline src/network/networkd-fdb.c]: If this fails, then skip to configure this entry. We should not configure incomplete entries.

**Review by yuwata — CHANGES_REQUESTED** [review]
In networkd-link.c, the list of alternative names are updated in several places. So, let's introduce, say `update_links_by_name()`.

File: src/network/networkd-link.c
Comment (ssahani) [reply src/network/networkd-link.c]: See https://github.com/systemd/systemd/pull/17248#discussion_r507828602

Comment (yuwata) [reply src/network/networkd-link.c]: Thank you for the reference. OK, let's keep this as is.

**Comment by ssahani** [comment]
Updated thanks for the review @yuwata ptal.

**Comment by ssahani** [comment]
Updated thanks for the review @yuwata ptal.

**Comment by ssahani** [comment]
@poettering ptal

File: man/systemd.network.xml
Comment (yuwata) [inline man/systemd.network.xml]: Maybe, `OutgoingInterface=`? In `[RoutingPolicyRule]` we already have `OutgoingInterface=`. So, I prefer `OutgoingInterface=` for consistency.

File: src/network/networkd-fdb.c
Comment (yuwata) [inline src/network/networkd-fdb.c]: I think we should skip to configure the entry rather than return failure when the device does not exist yet.
And, configure the entry when the corresponding interface comes.

Such a mechanism should also be used for `MultipathRoute=`, and not sure but maybe `OutgoingInterface=` and `IncomingInterface=` in `[RoutingPolicyRule]`.

File: src/network/networkd-fdb.c
Comment (yuwata) [inline src/network/networkd-fdb.c]: Please unset if an empty string is assigned.
```
if (isempty(rvalue)) {
        ftb_entry->ifname = mfree(fdb_entry->ifname);
        TAKE_PTR(fdb_entry);
        return 0;
}
```

File: src/network/networkd-fdb.h
Comment (yuwata) [inline src/network/networkd-fdb.h]: I prefer `outgoing_ifname`.

File: src/network/networkd-link.c
Comment (yuwata) [inline src/network/networkd-link.c]: I think these lines should be moved to `link_update_by_name()`.

File: src/network/networkd-link.h
Comment (yuwata) [inline src/network/networkd-link.h]: It seems `link_update_by_name()` can be static.

File: src/network/networkd-manager.c
Comment (yuwata) [inline src/network/networkd-manager.c:870]: This must be in the first commit.

File: src/network/networkd-manager.h
Comment (yuwata) [inline src/network/networkd-manager.h]: This must be in the first commit.

**Review by yuwata — CHANGES_REQUESTED** [review]
Please move the lines about `manager->links_by_name` to the first commit, and make the second commit contain only about fdb related changes.

File: src/network/networkd-fdb.c
Comment (ssahani) [reply src/network/networkd-fdb.c]: https://github.com/systemd/systemd/pull/17248#discussion_r510582939. Then why did you made this comment

File: src/network/networkd-fdb.h
Comment (ssahani) [reply src/network/networkd-fdb.h]: It does not matter

File: src/network/networkd-link.h
Comment (ssahani) [reply src/network/networkd-link.h]: It may be used somewhere later. Let's keep it that way.

File: src/network/networkd-link.c
Comment (ssahani) [reply src/network/networkd-link.c]: The allocation should he here .

**Comment by ssahani** [comment]
Cc @bluca

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: maybe move all this also into link_update_by_name(), so that that function comprehensively adds all names to the hashmap both the main one and the alternative ones, and implicitly creates the hashmap before, if it is missing

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: you are not actually ignoring here. I think ignoring is good, but please downgrade to `log_warning_errno()`, and drop the `return`

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: please move the name unregistration into a function of its own, too, also covering both the alternative names and the main name

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: if you have that unregistration function i suggested above, you could just call it here, simplifying things quite a bit

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c:2763]: the main name of an iface can chance too, and needs to be covered. if you properly encapsulate the update/removal functions as suggested, then you will get this for free, as one will remove the whole shebang including main name, and the other will register the whole shebang

File: src/network/networkd-manager.h
Comment (poettering) [inline src/network/networkd-manager.h]: maybe add a comment to both these tlines:

```c
        Hashmap *links; /* keyed by ifindex */
        Hashmap *links_by_name; /* keyed by main + alternative interface names */
```

**Review by poettering — CHANGES_REQUESTED** [review]
sorry for the late review. 

final review I figure we should leave to @yuwata

File: src/network/networkd-link.c
Comment (ssahani) [reply src/network/networkd-link.c:2763]: I did not got this.

**Comment by ssahani** [comment]
updated. Thanks for the review. I did not get the last point though

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: this check looks unnecessar, should work the same if these two lines are dropped

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: the conditionalization should go, no? this should be called in any case

File: src/network/networkd-link.c
Comment (poettering) [reply src/network/networkd-link.c:2763]: the main name of the interface can change anytime too, and we should cover that here too and update the by-name hash tables, too

**Review by poettering — CHANGES_REQUESTED** [review]
it would be great to split this in two commits btw: one that adds the by_name hash table, and the other one that then adds the fdb stuff based on it

**Comment by ssahani** [comment]
updated . Thanks for the review. ptal

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c:660]: maybe have two distinct functions link_remove_by_altnames() and link_remove_by_ifname() here, and then call either one or the other, or both. the boolean thing is a bit weird

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c:649]: hmm, why call this "update"? you use it to add only

I'd split this up into two functions: link_add_by_ifname() and link_update_by_altnames()... even if that means having the hashmap_ensure_allocated() in place twice

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: this doesn't look right, since you remove the entries for the new name here, but you should remove them for the *old* name

File: src/network/networkd-link.c
Comment (poettering) [inline src/network/networkd-link.c]: same here, you remove the entries for the new nmes, not the old names

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
the second commit looks good

File: src/network/networkd-link.c
Comment (ssahani) [reply src/network/networkd-link.c]: yeah right. I moved it up before link_drop.

**Comment by ssahani** [comment]
Updated. Thanks for the patience.

**Comment by awlx** [comment]
Any updates on this PR? That's also a feature we would need :).

**Comment by poettering** [comment]
PR looks good to me, but @yuwata should have a final review I guess. Also, needs a rebase.

**Pull request closed without merging**
